package/src/bin/thing-flinger.rs

```diff
@@ -0,0 +1,590 @@
+    restart_services(server)
+}
+
+// Copy package artifacts as a result of `omicron-package package` from the
+// builder to the deployment server staging directory.
+fn copy_package_artifacts_to_staging(
+    config: &Config,
+    pkg_dir: &str,
+    builder: &Server,
+    destination: &Server,
+) -> Result<()> {
+    let cmd = format!(
+        "rsync -avz -e 'ssh -o StrictHostKeyChecking=no' \
+                    --exclude overlay/ {} {}@{}:~/{}",
+        pkg_dir,
+        destination.username,
+        destination.addr,
+        config.deployment.staging_dir
+    );
+    println!("$ {}", cmd);
+    ssh_exec(builder, &cmd, true)
+}
+
+fn copy_omicron_package_binary_to_staging(
+    config: &Config,
+    builder: &Server,
+    destination: &Server,
+) -> Result<()> {
+    let mut bin_path = PathBuf::from(&config.builder.omicron_path);
+    bin_path.push("target/debug/omicron-package");
+    let cmd = format!(
+        "rsync -avz {} {}@{}:~/{}",
+        bin_path.to_string_lossy(),
+        destination.username,
+        destination.addr,
+        config.deployment.staging_dir
+    );
+    println!("$ {}", cmd);
+    ssh_exec(builder, &cmd, true)
+}
+
+fn copy_package_manifest_to_staging(
+    config: &Config,
+    builder: &Server,
+    destination: &Server,
+) -> Result<()> {
+    let mut path = PathBuf::from(&config.builder.omicron_path);
+    path.push("package-manifest.toml");
+    let cmd = format!(
+        "rsync {} {}@{}:~/{}",
+        path.to_string_lossy(),
+        destination.username,
+        destination.addr,
+        config.deployment.staging_dir
+    );
+    println!("$ {}", cmd);
+    ssh_exec(builder, &cmd, true)
+}
+
+fn run_omicron_package_from_staging(
+    config: &Config,
+    destination: &Server,
+    artifact_dir: &Path,
+    install_dir: &Path,
+) -> Result<()> {
+    let mut deployment_src = PathBuf::from(&config.deployment.staging_dir);
+    deployment_src.push(&artifact_dir);
+
+    // Run `omicron-package install` on the deployment server
+    let cmd = format!(
+        "cd ~/{} && pfexec ./omicron-package install --in ~/{} --out {}",
+        config.deployment.staging_dir,
+        deployment_src.to_string_lossy(),
+        install_dir.to_string_lossy()
+    );
+    println!("$ {}", cmd);
+    ssh_exec(destination, &cmd, true)
+}
+
+fn copy_overlay_files_to_staging(
+    config: &Config,
+    pkg_dir: &str,
+    builder: &Server,
+    destination: &Server,
+    destination_name: &str,
+) -> Result<()> {
+    let cmd = format!(
+        "rsync -avz {}/overlay/{}/ {}@{}:~/{}/overlay/",
+        pkg_dir,
+        destination_name,
+        destination.username,
+        destination.addr,
+        config.deployment.staging_dir
+    );
+    println!("$ {}", cmd);
+    ssh_exec(builder, &cmd, true)
+}
+
+fn install_overlay_files_from_staging(
+    config: &Config,
+    destination: &Server,
+    install_dir: &Path,
+) -> Result<()> {
+    let cmd = format!(
+        "pfexec cp -r ~/{}/overlay/* {}",
```

Comment:
Does this `~` need to be the username for this server from the config?

Reply:
As far as I can tell ssh with both bash and zsh do the right thing on illumos for both types of shell modes. 

The reason I did it this way was to reduce the amount of configuration necessary for each server. The path doesn't really matter, we just want it to be the home directory. I could also use $HOME, but this appears to work fine.

Reply:
On second thought, I am just going to change to `$HOME` to remove one more potential problem.

Reply:
Done.